### ISSUE TITLE
Multiple CONNECTION_SUCCESS responses for ConnectionStateCallbacks

### ISSUE THREAD

**Issue by rkanubaddhi** [Azure/azure-iot-sdk-java#217]
- **OS and version used: Windows 10** <VERSION> 

- **Java runtime used: Java 8** <VERSION> 

- **SDK version used: 1.9.0** <VERSION> 

# Description of the issue:

I'm opening a connection using deviceClient.open(), before opening the connection I'm registering the connection callbacks. execute method of IotHubConnectionStateCallback is getting executed multiple times for single open() and I'm getting IotHubConnectionState.CONNECTION_SUCCESS for 7 times in the multiple execute calls. Please look at the code below.

# Code sample exhibiting the issue:

```
public void connect() throws URISyntaxException
  {
    dc = new DeviceClient(deviceconnectionString, IotHubClientProtocol.AMQPS_WS);
    try
    {
      dc.registerConnectionStateCallback(new IotHubConnectionStateCallback()
         {
           @Override
           public void execute(IotHubConnectionState msgCode, Object context)
           {
             if(msgCode == IotHubConnectionState.CONNECTION_SUCCESS)
              System.out.println("Connection Success:" +msgCode);
           }
         }, null);
      dc.open();
    }
    catch (IOException e)
    {
      e.printStackTrace();
    }
}
```

 

# Console log of the issue:

I see the connection success log 7 times.

Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS
Connection Success:CONNECTION_SUCCESS

**Comment by timtay-microsoft** [comment]
We are currently working on a bit of an overhaul to the Connection Status reporting. In the next release, the connection status callback API you are using will be deprecated and replaced by a new API. That new API will provide the fix you need for this scenario. I'll let you know when that release occurs so that you can test it out.

**Comment by rkanubaddhi** [comment]
Thank you for the quick reply, is there any ETA?

**Comment by timtay-microsoft** [comment]
We're shooting to have this released on or before 4/6. I'll update this thread as soon as we release.

**Comment by rkanubaddhi** [comment]
Verified this in 1.10.0 and this has been fixed. Thank you for the quick fix!

**Comment by rkanubaddhi** [comment]
Initially on a connection I only saw 1 connection callback, however, if I have messages in the queue while I'm disconnected, I didn't receive CONNECTION_DROP callback. Once I come back on connection. I see multiple CONNECTION_SUCCESS callback
Connection Success:CONNECTION_SUCCESS 
Sending message....:0                 
Sending message....:1                 
Sending message....:2                 
Sending message....:3                 
Sending message....:4                 
Sending message....:5                 
Sending message....:6                 
Sending message....:7                 
Sending message....:8                 
Sending message....:9                 
Message sent successfully!!!:6        
Message sent successfully!!!:8        
Message sent successfully!!!:7        
Message sent successfully!!!:9        
Message sent successfully!!!:3        
Message sent successfully!!!:4        
Message sent successfully!!!:5        
Message sent successfully!!!:1        
Message sent successfully!!!:0        
Message sent successfully!!!:2        
Sending message....:0                 
Sending message....:1                 
Sending message....:2                 
Sending message....:3                 
Sending message....:4                 
Sending message....:5                 
Sending message....:6                 
Sending message....:7                 
Sending message....:8                 
Sending message....:9                 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
Connection Success:CONNECTION_SUCCESS 
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED

This is what I have:

`dc.registerConnectionStateCallback(new IotHubConnectionStateCallback()
         {
           @Override
           public void execute(IotHubConnectionState msgCode, Object context)
           {
             if(msgCode == IotHubConnectionState.CONNECTION_SUCCESS)
              System.out.println("Connection Success:" +msgCode);
             if(msgCode == IotHubConnectionState.CONNECTION_DROP)
               System.out.println("Connection drop:" +msgCode);
           }
         }, null);`

**Comment by rkanubaddhi** [comment]
I have also tried with new ConnectStatusCallback API:

Here is the observation:
Sent 10 messages, dropped the network connection I saw following:
Sending message....:0
Sending message....:1
Sending message....:2
Sending message....:3
Sending message....:4
Sending message....:5
Sending message....:6
Sending message....:7
Sending message....:8
Sending message....:9
Connection drop, retrying:DISCONNECTED_RETRYING

/////////////////////////////////////////////////////////////
Waited for couple of minutes without connection then connected the device back. Here is what I saw:
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
failed to send message:MESSAGE_EXPIRED
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED
failed to send message:MESSAGE_EXPIRED

Whereas what I'm expecting is single connection callback, once the device comes back online. am I missing something? are there any examples on how to consume the new API?

Here is my code:
```
dc.registerConnectionStatusChangeCallback(new IotHubConnectionStatusChangeCallback()
      {
        @Override
        public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext)
        {
          if(status == IotHubConnectionStatus.CONNECTED)
            System.out.println("Connection Success:" +status);
          if(status == IotHubConnectionStatus.DISCONNECTED)
            System.out.println("Connection drop:" +status);
          if(status == IotHubConnectionStatus.DISCONNECTED_RETRYING)
            System.out.println("Connection drop, retrying:" +status);
        }
      }, null);
```

**Comment by timtay-microsoft** [comment]
I highly recommend you disregard the deprecated API for connection status updates. Since it is deprecated, we won't be supporting it anymore. 

Focusing on the new registerConnectionStatusChangeCallback, are you seeing all of those CONNECTED and DISCONNECTED_RETRYING callbacks fired at the same time, or is there some gap that separates each callback? Are you seeing this same behavior from MQTT/MQTT_WS as well? What about AMQPS?

It would also be helpful to have the proton-j logs of what the amqp connection looks like during all of this. If you set the environment variable "PN_TRACE_FRM" to the value 1 you can get these logs. If you attach those logs, It would be much easier to fix this issue

**Comment by rkanubaddhi** [comment]
I have tried only AMQPS over WS. I have seen all those CONNECTED and DISCONNECTED_RETRYING fired at the same time immediately upon reconnection. Sure I will attach the logs today.

**Comment by timtay-microsoft** [comment]
Also, are you multiplexing (Transport Client)? Or are you just using the device client?

**Comment by rkanubaddhi** [comment]
just using device client

**Comment by rkanubaddhi** [comment]
[logs.txt](https://github.com/Azure/azure-iot-sdk-java/files/1921159/logs.txt)

**Comment by timtay-microsoft** [comment]
You are using the new API correctly. From what I can tell, your problem is your internet connection. It seems like the TCP connection is being dropped as soon as it is established. Maybe some sort of firewall issue?

I can't replicate the behavior seen in your logs. How consistently do you see this behavior? Can you try using MQTT just for comparison's sake?

**Comment by timtay-microsoft** [comment]
@rkanubaddhi are you still seeing this behavior? Because I am still unable to repro

**Comment by timtay-microsoft** [comment]
Okay, I was able to repro your current issue and have found the fix. I'll make sure to get this fix finalized before the next release we do.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been checked in and should be released tomorrow.

**Comment by rkanubaddhi** [comment]
Thanks for the quick response!

**Comment by rkanubaddhi** [comment]
I have tested it with 1.11.0, and looks like problem still exists. Please let me know if you are not able to reproduce it.

**First iteration:**
Connection Success:CONNECTED
Sending message....:0
Sending message....:1
Sending message....:2
Sending message....:3
Sending message....:4
Message sent successfully!!!:0
Message sent successfully!!!:1
Message sent successfully!!!:2
Message sent successfully!!!:3
Message sent successfully!!!:4

**Second iteration** //disconnected the network immediately after sending the 5 messages and connected it back after receiving _"Connection drop, retrying:DISCONNECTED_RETRYING"_
Sending message....:0
Sending message....:1
Sending message....:2
Sending message....:3
Sending message....:4
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Message sent successfully!!!:4
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Connection drop, retrying:DISCONNECTED_RETRYING
Connection Success:CONNECTED
Message sent successfully!!!:1
Message sent successfully!!!:2
Message sent successfully!!!:3
Message sent successfully!!!:0

**Comment by timtay-microsoft** [comment]
Looks like you are right, this issue does sometimes still pop up. I'll keep digging and keep you posted.

**Comment by timtay-microsoft** [comment]
I have checked in the fix to this issue as of 23d7c90125a9ee5e35fa3e527812b89e95ef1bd3.

Feel free to clone our repo, build from source, and try it out!

It will be released formally 5/25

**Comment by rkanubaddhi** [comment]
Sure, I will try it out today, thanks!

**Comment by rkanubaddhi** [comment]
I have built from the master branch and tried it out, it looks good now. I will test it again once it is released on 5/25. 

Thank you!

**Comment by rkanubaddhi** [comment]
I have tested it building from the repo and looks like it has been fixed. I will test it again once it is released.
Thanks!

**Comment by timtay-microsoft** [comment]
Glad to hear it! Thanks for being so vigorous with looking into this edge case. It really helps the SDK.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been released as of Device Client 1.11.1

Since this issue has been fixed, I'm closing this issue. Feel free to reopen if you still see this issue.

@rkanubaddhi thank you for your contribution to our open-sourced project!  Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V).
